shopinvader_address/models/res_partner.py

```diff
@@ -0,0 +1,162 @@
+
+    def _get_shopinvader_billing_address(self) -> "ResPartner":
+        self.ensure_one()
+        return self
+
+    def _create_shopinvader_billing_address(self, vals: dict) -> "ResPartner":
+        raise UserError(_("Creation of billing addresses is not supported"))
+
+    def _update_shopinvader_billing_address(
+        self, vals: dict, address_id: int
+    ) -> "ResPartner":
+        self.ensure_one()
+
+        if address_id != self.id:
+            raise UserError(
+                _(
+                    "Can not update billing addresses(%(address_id)d)"
+                    "because it doesn't correspond to authenticated partner's billing address",
+                    address_id=self.id,
+                )
+            )
+
+        # if billing addresses is already used, it is not possible to modify it
+        if self._shopinvader_billing_address_already_used():
+            raise UserError(
+                _(
+                    "Can not update billing addresses(%(address_id)d)"
+                    "because it is already used on confirmed sale order",
+                    address_id=self.id,
+                )
+            )
+
+        self.write(vals)
+
+        return self
+
+    # ---Shipping ---
+
+    def _get_shopinvader_shipping_addresses(
+        self, address_id: int = None
+    ) -> "ResPartner":
+        self.ensure_one()
+        domain = [("type", "=", "delivery"), ("parent_id", "=", self.id)]
+
+        if address_id is not None:
+            domain.append(("id", "=", address_id))
+
+        return self.env["res.partner"].search(domain)
+
+    def _create_shopinvader_shipping_address(self, vals: dict) -> "ResPartner":
+        self.ensure_one()
+        vals.update(
+            {
+                "parent_id": self.id,
+                "type": "delivery",
+            }
+        )
+        return self.env["res.partner"].create(vals)
+
+    def _update_shopinvader_shipping_address(
+        self, vals: dict, address_id: int
+    ) -> "ResPartner":
+
+        if any(key in vals for key in ("parent_id", "type")):
+            raise UserError(
+                _(
+                    "parent_id and type cannot be modified on"
+                    " shopinvader shipping address, id: %(address_id)d",
+                    address_id=address_id,
+                )
+            )
+
+        self.ensure_one()
+        address = self._get_shopinvader_shipping_addresses(address_id)
+        if not address:
+            raise MissingError(
+                _("No address found, id: %(address_id)d", address_id=address_id)
+            )
+
+        # if shipping addresses is already used, it is not possible to modify it
+        if address._shopinvader_shipping_address_already_used():
+            raise UserError(
+                _(
+                    "Can not update Shipping address(%(address_id)d)"
+                    "because it is already used on confirmed sale order",
```

Comment:
```suggestion
                    "because it is already used.",
```

Strictly speaking, in this function, we don't know what `_shopinvader_shipping_address_already_used()` does exactly, so we can't be too precise in the error message.

Alternatively if we want a more precise error message, we could rename `_shopinvader_shipping_address_already_used` to `_ensure_shopinvader_shipping_address_not_used` and let that one raise the exception since it knows what it tests precisely.